### ISSUE TITLE
Require external (unmanaged) file

### ISSUE THREAD

**Issue by kaste** [webpack/webpack#150]
I would have expected that a

```
require.ensure(['script!//ajax.googleapis.com/ajax/libs/angularjs/1.2.9/angular.min.js'], function(require){

});
```

would work out-of-the-box. But it doesn't.

(I don't need this for angular, but would like to do this for some payments-gateway-js.)

Any solution for this.

**Comment by jbcpollak** [comment]
I was just looking for something like this but for fonts, not scripts. It doesn't seem common fonts are stored in npm, so rather than maintaining my own variants I was hoping I could convince webpack to download them.

@m59peacemaker - It looks like you removed your loader project, do you still have the source?

**Comment by m59peacemaker** [comment]
@jbcpollak I haven't used webpack in a long time, so I don't know how fonts should work. My loader was bad and wrong, so I removed it.

**Comment by simkessy** [comment]
Would it be possible to do something similar but with CSS. I'm assuming $scriptjs would only work with js files.

**Comment by jbcpollak** [comment]
@simkessy - Thats similar to what I'm looking for (for fonts) - I can imagine the same for images, etc. It seems like a useful feature for webpack.

**Comment by simkessy** [comment]
Yea it would be nice. If you find anything let me know.

**Comment by mc-zone** [comment]
> webpack is a module bundler not a javascript loader. It package files from local disk and don't load files from the web (except its own chunks).
> 
> Use a javascript loader, i. e. script.js

And how it can be work if we want to load a extenal-script which have been written with module API(for example AMD) ?
In other word, is there any way we can add a module into webpack module list at runtime ?  :mask:

**Comment by jhnns** [comment]
@mc-zone you should be able to load AMD modules with script.js. But this has nothing to do with webpack, it just works because AMD was designed for this. Thus you won't be able to use webpack features inside these AMD modules. webpack requires static analysis on build time.

If you want to include another webpack build, you could try to use the [DllPlugin](http://webpack.github.io/docs/list-of-plugins.html#dllplugin) to compile the library and the [DllReferencePlugin](http://webpack.github.io/docs/list-of-plugins.html#dllreferenceplugin) to reference this library inside your bundle without including it.

**Comment by oveddan** [comment]
We have a similar issue.  We have large video files that would be too big to check into our local code repository - but are loaded via a cdn.  Is there an easy way to tell webpack to make the path for these files come from a specific external host?

**Comment by jhnns** [comment]
@oveddan Why would you want to manage that file with webpack? What would be the expected behavior?

**Comment by jbcpollak** [comment]
As I mentioned before, this behavior would be handy for things like fonts that you want embedded in your application but are only available as URLs online. 

It doesn't seem appropriate to have to download fonts and check them into source control when you could refer to them by URL.

I realise you could refer to them by URL at runtime, but I would rather webpack bundle them for us. (And in my case I am building an offline app so I need the fonts local to the application)

**Comment by LukasBombach** [comment]
I could second that this would be handy. We have to lazy load an external script that we do not want to bundle since our vendor might update it at will. using webpack as a single source management solution would be awesome. A loader would be nice for this, maybe something similar to `require("bundle?lazy!./file.js")` where instead of `./file.js` we could have `//yadda.something.you/know.js`.

**Comment by sergeymorkovkin** [comment]
My preference would probably be:

1. In externals we specify CDN URL starting from //
2. Webpack adds corresponding <script> tags to html

**Comment by kevinwheeler** [comment]
If the script I load using script.js defines a global variable that I need to use, how would I then use this global variable?

```
var $script = require("scriptjs");
$script("http://example.com/someScript.js", function() {
  // how do I use the global variable here that someScript defined?
});
```

**Comment by jhnns** [comment]
`window.globalVariable` or even just  `globalVariable`? 😁

**Comment by kevinwheeler** [comment]
Oh, lol, sorry. This doesn't work for me because I am doing this inside an es6 module (using babel), so no access to global variables.

**Comment by jhnns** [comment]
You can still access global variables inside an es6 module

**Comment by shachar-i** [comment]
I'm completely new to Webpack, but this makes no sense to me.
I am assuming that I'm missing something here, but could not figure it out.

The fact Webpack has no (built-in, easy) way to resolve remote dependencies discourages CDN usage for popular packages like React, jQuery etc.. (I realize there are ways around it, but none are pretty as far as I could find and it is seems like the less common usage).

I just started using Webpack and the most natural thing to do was `npm install` everything I needed.
But then my bundle quickly became huge, and cached versions of these libs I installed would no longer be used to speed up loading of my app.

So is Webpack's design actually encouraging bloated bundles (with many included libs), and slow-loading web apps??? 
I must be missing something...

**Comment by nkbt** [comment]
@shachar-i please, check this out https://webpack.github.io/docs/library-and-externals.html

**Comment by nkbt** [comment]
You can use different configs for dev/prod. Dev without externals, so all scripts go into bundle, prod with - so files are used from CDN.

**Comment by shachar-i** [comment]
Thanks @nkbt , but if I use externals, I'll have to take care of my own dependencies (which is a major reason to start using Webpack to begin with).
This is actually what I did (use externals), but not the point I'm trying to make.

Since bundling npm installed packages is way nicer and more _Webpacky_, **Isn't Webpack encouraging bloated bundles by not giving a good remote CDN option for 3rd party libs???**
A half-lazy developer would just pick the easy big-bundle way. I supposed many did.
For example, I've looked at several react-webpack tutorials. All of them encourage 'npm install react' rather than the ugly external.

**Comment by nkbt** [comment]
I understand where you are going here, but the thing is - CDN is an app concern, not bundler. Besides, there is no such thing as `good remote CDN option for 3rd party libs` except maybe the recent `unpkg` (npmcdn). But for any reasonably loaded prod it makes sense to buy/setup your own CDN.

I personally cannot come up with any reasonable answer to this. If you do - PR to docs would be a good point to start the discussion.

PS: mastering your tools is always a difficult process that takes heaps of time and errors. If it is possible to reduce this time - awesome

**Comment by mnpenner** [comment]
I haven't thoroughly tested it yet, but you can use something like this to load an external script:

```js
const _scriptCache = new Map();

/**
 * @param {string} url
 * @param {string=} name
 * @return {Promise}
 */
export function loadScript(url, name) {
    let promise;

    if(_scriptCache.has(url)) {
        // TODO: normalize URL
        promise = _scriptCache.get(url);
    } else {
        promise = new Promise((resolve,reject) => {
            let script = document.createElement('script');
            script.onerror = event => reject(new Error(`Failed to load '${url}'`));
            script.onload = resolve;
            script.async = true;
            script.src = url;

            if(document.currentScript) {
                document.currentScript.parentNode.insertBefore(script, document.currentScript);
            } else {
                (document.head || document.getElementsByTagName('head')[0]).appendChild(script);
            }
        });

        _scriptCache.set(url, promise);
    }

    return promise.then(() => {
        if(global[name]) {
            return global[name];
        } else {
            throw new Error(`"${name}" was not created by "${url}"`);
        }
    });
}
```

Usage:

```js
loadScript('https://js.stripe.com/v2/', 'Stripe').then(Stripe => {
    console.log('Stripe loaded!', Stripe);
});
```

The `name` parameter is to ensure the API you're loading actually created the global variable you were expecting (e.g. Stripe creates `window.Stripe`) and it avoids some IDE warnings ("Stripe is undefined").

**Comment by ryancwalsh** [comment]
I totally agree with @shachar-i 's points above at https://github.com/webpack/webpack/issues/150#issuecomment-257114083. 

I have spent 8 hours trying to learn Webpack today, and the docs do not clarify:

I want to know how a Webpack website application (with multiple entry points) can manage lots of different dependencies (which I thought was the main motivation for Webpack) while also taking advantage of the fact that many people have jQuery and other popular libraries cached in their browsers because they've already visited sites that pulled them from CDNs, such as https://ajax.googleapis.com/ajax/libs/jquery/2.2.3/jquery.min.js.

Why would I serve my own copies of popular libraries (e.g. bundled into the rest of my JS)? That seems like a waste of browser payload. 

But I also don't yet understand @sokra 's https://github.com/webpack/webpack/issues/150#issuecomment-32756166 recommending scriptJs.

Why can't Webpack manage them CDN dependencies?

What am I misunderstanding?

Thanks, everyone!

**Comment by mnpenner** [comment]
@ryancwalsh 

> I want to know how a Webpack website application (with multiple entry points) can manage lots of different dependencies ... while also taking advantage of the fact that many people have jQuery and other popular libraries cached in their browsers because they've already visited sites that pulled them from CDNs

You can do this the same way you always would. Add

```js
<script src="https://code.jquery.com/jquery-3.2.1.min.js" integrity="sha384-xBuQ/xzmlsLoJpyjoggmTEz8OWUFM0/RC5BsqQBDX2v5cMvDHcMakNTNrHIW2I5f" crossorigin="anonymous"></script>
```

To the top of your page, and then use `$` or `window.$` in your bundle.

> Why would I serve my own copies of popular libraries (e.g. bundled into the rest of my JS)? That seems like a waste of browser payload.

It's your choice. If you trust your CDN to always be available, don't bundle it. I personally put all my vendor/3rd party libraries in a [CommonsChunk](https://webpack.js.org/plugins/commons-chunk-plugin/) and perma-cache it. Our users only have to download this bundle once the first time they visit the site. Since it's a web-app, 99% of our users are recurring. If most of your users are new visitors, you might prefer a CDN.

> But I also don't yet understand @sokra 's #150 (comment) recommending scriptJs.  
> Why can't Webpack manage them CDN dependencies?

Because it's not webpack's job to do that for you. Its purpose is to bundle local files, not manage CDNs. It's relatively easy to add a `<script src="...">` to the top of your page, use `scriptjs`, or create a helper function like I did. There's no need to bake that into webpack.

What would that even look like? Supporting `require('https://...')` would lock up your browser while the script downloads, because `require` is synchronous. `require.ensure` is async and support *could* be added there, but it's just as trivial to create your own `myExternalRequire`.

**Comment by ryancwalsh** [comment]
@mnpenner Thanks for the response!!  ☺ 

Yes, most of my site's visitors are new and not recurring.  Maybe I'm in the minority and starting to use Webpack was the wrong choice for me.

I'm surprised more Webpack fans aren't asking for this functionality. To me, managing dependencies is important, regardless of whether the code is hosted all together locally or if some of it sits on a CDN.

I'm not sure how to create my own `myExternalRequire` function to do what we're describing. 

[I'm impressed that I even got Webpack working at all—it took me more than a full day to migrate my site over (to the Webpack way). The docs are super sparse and seem like they were written as reminders to the creators of Webpack rather than to introduce new developers to it.]

**Comment by mnpenner** [comment]
@ryancwalsh I posted my implementation just [a few comments up](https://github.com/webpack/webpack/issues/150#issuecomment-275952354). Webpack *is* difficult to get started with, I'll grant you that. I'm sure the maintainers wouldn't mind some help on the documentation side once you get it all figured out. There's lots of blogs/tutorials on the subject if you find the official documentation too sparse.

**Comment by ryancwalsh** [comment]
@mnpenner Sweet--I don't know how I missed that. I'll check that out. Thanks!

**Comment by bizm** [comment]
Is it possible to bundle library as a module for dynamic import (https://webpack.js.org/guides/code-splitting-async/#dynamic-import-import-)?

I just want my library code to be wrapped by webpackJsonp function (without all that webpack bootstrap) and then in a parent project i want to have import("path/to/external/library.js") which doesn't throw any "Module not found" errors.

**Comment by wbern** [comment]
In my use case, I want to be able to use System.import() to import some external url's that I only want resolved during runtime.

After scouring this page a bit: https://github.com/webpack/docs/wiki/configuration#automatically-created-contexts-defaults-modulexxxcontextxxx

The simplest thing I got working was to disable the error in the webpack config.

```
module: {
        exprContextCritical: false, // true triggers warning
        rules: [ ...
```

I tried to figure out how to disable the warning only for certain imports, like if the expression started with certain letters like "url", or if I had provided a string literal that said "http", using the remaining properties. I couldn't figure it out unfortunately. If anyone else knows how, would be happy to know!

**Comment by wbern** [comment]
Ok, I found how to import a webpack bundle from another project.

```
fetch(appUrl).then(response => response.text()).then(function(text) {
    let theModule = null;
    let define = (a) => theModule = a();
    eval(text); // the imported module ends up in theModule
});
```

**Comment by pavel-agarkov** [comment]
Hello.
I have some pre-bundled with webpack angular modules stored on the another server.
I want to lazy-load them when it is necessary from both client and server side (server pre-rendering).
I have tried already:
1. **$script**js => but it uses _document_ so it doesn't work on the server side
2. **require**js => looks like it overlapping with webpack _require_ method that causes an error: 
> ERROR in ./~/requirejs/bin/r.js
>Module parse failed: /app/node_modules/requirejs/bin/r.js Unexpected character '#' (1:0)
>You may need an appropriate loader to handle this file type.

Is there any other way to lazy load external js-bundles that does work with angular 4 and webpack?

**Comment by MrFrankel** [comment]
I've written a loader that allows you to expose modules from one webpack build to others via a global defined namespace, this lets you define which modules you want to share between the webpack builds and allows you to share common modules while maintaining the bundles intact.

https://github.com/MrFrankel/share-loader/blob/master/README.md

**Comment by yvele** [comment]
A possible workaround my be using `__non_webpack_require__ ` that is is the original `require` function which exists in the environment before webpack comes in. See https://github.com/webpack/webpack/issues/1021#issuecomment-97440458

**Comment by josiahruddell** [comment]
I posted a workaround here: https://stackoverflow.com/a/54559637/39013

This solution uses webpack's `noParse` configuration to make a module that exports the native `require` function to bundled code.

**Comment by pimterry** [comment]
It looks like there's a few different cases being discussed here, but if you're trying to require an external file and pull it in at _build_ time (e.g. pull it from a CDN and include it in your bundle), I wrote a blog post on how to do that using val loader: https://httptoolkit.tech/blog/bundling-remote-scripts-with-webpack/

In short: you can write a tiny script and load it with [val loader](https://github.com/webpack-contrib/val-loader) to make webpack read the latest content from the CDN on every build and include it in your bundle, so you can treat it like any other normal dependency without having to vendor it into your repo.
